### DISCUSSION TITLE
Where to put code to artificially slow down min.io requests?

### DISCUSSION THREAD

**Discussion by celan69** [minio/minio#11672] [Q&A] [answered]
Hi all,
my public-sector employer is introducing an electronic filing system using an S3-compatible large-scale storage backend. I have no idea from which vendor (different department). This S3-compatible storage backend takes between 300 and 1000ms to serve an object request. In conjuction with some possible implementation and/or architectural issues in the e-filing system, this leads to poor performance for the end user.

To troubleshoot these issues I would like to use min.io in my development enviroment and make it answer incoming requests equally slowly.

Where in the source code would be a good location to put in some artificial delays?

Sorry for this highly unconventional request ;-)

Regards and thank you for min.io,
Christian

**Comment by harshavardhana** [comment, marked as answer]
You should do this outside MinIO that use `tc` to do traffic control 

```
# Refer to: http://www.linuxfoundation.org/collaborate/workgroups/networking/netem#Delaying_only_some_traffic
# Refer to: http://www.bomisofmab.com/blog/?p=100
# Refer: http://drija.com/linux/41983/simulating-a-low-bandwidth-high-latency-network-connection-on-linux/
# Setup the rate control and delay
sudo tc qdisc add dev lo root handle 1: htb default 12 
sudo tc class add dev lo parent 1:1 classid 1:12 htb rate 56kbps ceil 128kbps 
sudo tc qdisc add dev lo parent 1:12 netem delay 1000ms

# Remove the rate control/delay
sudo tc qdisc del dev lo root

# To see what is configured on an interface, do this
sudo tc -s qdisc ls dev lo

# Replace lo with eth0/wlan0 to limit speed from LAN
```

**Comment by celan69** [comment]
Sorry, forgot to mention that the whole development environment I was talking about is running on Windows (in a virtual Machine).

Thanks anyway, I will try to figure something out!

**Reply by klauspost** [reply]
You could look into a proxy that offers this functionality. I wouldn't be surprised if that existed.

**Comment by celan69** [comment]
I have put in a "time.Sleep(1*time.Second)" in GetObjectHandler and PutObjectHandler in cmd/object-handlers.go.

Seems to do the (admittedly, crude) job for the time being :-)
